### PR TITLE
Improve handling and documentation about supported and unsupported value types in droption

### PR DESCRIPTION
**Pull request by edmcman**
- Add default implementations of convert_from_string and default_as_string to avoid crashes.
- Add implementations for long and long long value types
- Update documentation to list the supported value types

### CODE REVIEW

**Comment by AssadHashmi** [comment]
run arm tests

**Comment by AssadHashmi** [comment]
@edmcman I've manually kicked off the AArch64 precommit tests and added you to the Jenkins builder ACL. From now on AArch64 precommits should be automatically triggered when you commit.

**Comment by edmcman** [comment]
It looks like Travis was failing because of a pre-C++11 build.  In pre-C++11, I added an assertion failure, but maybe there is a DR-specific way that this should be handled.

**Comment by derekbruening** [comment]
> It looks like Travis was failing because of a pre-C++11 build. In pre-C++11, I added an assertion failure, but maybe there is a DR-specific way that this should be handled.

All the internal builds are C++11.  The proj_samples test is a test of using the public API as an external user would.  Since we do not require all users of the API to use C++11 it is not built that way.  We may eventually do that but it would be a backward compatibility break.  If you think that would provide multiple benefits please file an issue and maybe we could do that sooner.

File: ext/droption/droption.h [hunk 1]
```diff
@@ -539,6 +540,20 @@ droption_t<int>::convert_from_string(const std::string s)
 }
 template <>
 inline bool
+droption_t<long>::convert_from_string(const std::string s)
+{
+    value = atol(s.c_str());
```
Comment (derekbruening) [inline ext/droption/droption.h]: Should we use strtol so we get error detection?  Ditto below (and in the original code looks like).

**Comment by edmcman** [comment]
> If you think that would provide multiple benefits please file an issue and maybe we could do that sooner.

I don't think changing this is necessary for this sake.  I just wasn't sure how I should handle the runtime error.  Should I do something beyond calling `assert`?  E.g., is there a `dr_report_error` function or something similar?

File: suite/tests/CMakeLists.txt [hunk 2]
```diff
@@ -2170,7 +2170,7 @@ if (CLIENT_INTERFACE)
     "" "" "")
 
   tobuild_ci(client.option_parse client-interface/option_parse.cpp
-    "-x;4;-y;quoted string;-z;first;-z;single quotes -dash --dashes;-front;value;-y;accum;-front2;value2;-no_flag;-takes2;1_of_4;2_of_4;-takes2;3_of_4;4_of_4;-val_sep;v1.1 v1.2;-val_sep;v2.1 v2.2;-val_sep2;v1;v2;-val_sep2;v3;v4;-large_bytesize;9999999999"
+    "-l;-4;-ll;-749882561;-ul;4;-ull;823636799;-x;4;-y;quoted string;-z;first;-z;single quotes -dash --dashes;-front;value;-y;accum;-front2;value2;-no_flag;-takes2;1_of_4;2_of_4;-takes2;3_of_4;4_of_4;-val_sep;v1.1 v1.2;-val_sep;v2.1 v2.2;-val_sep2;v1;v2;-val_sep2;v3;v4;-large_bytesize;9999999999"
```
Comment (derekbruening) [inline suite/tests/CMakeLists.txt]: It seems best to pass values that only a long long could hold.  These are both under 1 billion and so a regular long (at least for ILP64) will hold it.  I would add at least one more digit.

File: ext/droption/droption.h [hunk 3]
```diff
@@ -600,12 +641,34 @@ droption_t<twostring_t>::convert_from_string(const std::string s)
...
 }
+
+template <typename T>
+inline std::string
+droption_t<T>::default_as_string() const
+{
+#if __cplusplus >= 201103L
+    static_assert(sizeof(T) == -1, "Use of unsupported droption_t type");
+#else
+    assert(false && "Use of unsupported droption_t type");
```
Comment (derekbruening) [inline ext/droption/droption.h]: DR_ASSERT_MSG (and no need for assert.h) may be more suitable.

**Comment by edmcman** [comment]
AppVeyor iscomplaining about `code_api|tool.drcachesim.coherence` timing out.  Is this just a false positive?

**Comment by johnfxgalea** [comment]
@edmcman Yes, if I am not mistaken this is related to issue #3320.

**Comment by edmcman** [comment]
In that case I think this is ready to be merged.

**Comment by derekbruening** [comment]
> @edmcman Yes, if I am not mistaken this is related to issue #3320.

No this is new, but looks like a slow test that happened to hit the timeout limit.  I filed #3971.  Hopefully someone will have time to go tweak the test or timeout.

**Pull request merged**